pkg/controllers/namespace/subnet_share_controller.go

```diff
@@ -0,0 +1,244 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+// createSubnetCRInK8s creates the Subnet CR in Kubernetes
+func (r *NamespaceReconciler) createSubnetCRInK8s(ctx context.Context, subnetCR *v1alpha1.Subnet, subnetName string) error {
+	err := r.Client.Create(ctx, subnetCR)
+	if err != nil {
+		// If the Subnet CR already exists with the same name, try using generateName
+		if apierrors.IsAlreadyExists(err) {
+			log.Info("Subnet CR with the same name already exists, using generateName",
+				"Namespace", subnetCR.Namespace, "Name", subnetName)
+
+			// Create a new Subnet CR with generateName
+			// subnetCR.ObjectMeta.Name will be subnetName + "-" + randomSuffix
+			subnetCR.ObjectMeta.Name = ""
+			subnetCR.ObjectMeta.GenerateName = subnetName + "-"
+
+			err = r.Client.Create(ctx, subnetCR)
+			if err != nil {
+				return fmt.Errorf("failed to create Subnet CR with generateName: %w", err)
+			}
+		} else {
+			log.Error(err, "Failed to create Subnet CR", "Namespace", subnetCR.Namespace, "Name", subnetName)
+			return fmt.Errorf("failed to create Subnet CR: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// createSharedSubnetCR creates a Subnet CR for a shared subnet
+func (r *NamespaceReconciler) createSharedSubnetCR(ctx context.Context, ns string, sharedSubnetPath string) error {
+	// Extract the org id, project id, VPC id, and subnet id
+	orgID, projectID, vpcID, subnetID, err := common.ExtractSubnetPath(sharedSubnetPath)
+	if err != nil {
+		return err
+	}
+
+	vpcFullName, err := servicecommon.GetVPCFullName(orgID, projectID, vpcID, r.VPCService)
+	if err != nil {
+		return err
+	}
+
+	// Get associated resource name
+	associatedName, err := common.ConvertSubnetPathToAssociatedResource(sharedSubnetPath)
+	if err != nil {
+		return err
+	}
+
+	// Create the Subnet CR object
+	subnetCR := r.SubnetService.BuildSubnetCR(ns, subnetID, vpcFullName, associatedName)
+
+	// Create the Subnet CR in Kubernetes
+	err = r.createSubnetCRInK8s(ctx, subnetCR, subnetID)
```

Comment:
since the nsx subnet name is already in Subnet CR, no need to pass subnet name.

Reply:
fixed